File: java/goodmem-semantic-kernel/src/test/java/ai/goodmem/semantickernel/RecordingHttpServer.java
package ai.goodmem.semantickernel;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A local HTTP server on a real socket. It records the request line of every
 * request exactly as it arrived on the wire, and answers like a minimal GoodMem
 * server with one space named "notes".
 *
 * <p>A raw socket rather than WireMock or the JDK server, so nothing between the
 * connector and the recording normalises or rejects the path.
 */
final class RecordingHttpServer implements AutoCloseable {

    static final String DEFAULT_SPACE_ID = "01a0d16b-bbcd-701c-bfb4-fa306021e078";
    static final String ASSIGNED_MEMORY_ID = "019cfd1d-5a1e-7a4b-9c3e-2f6a1b0c0e09";

    private static final ObjectMapper MAPPER = new ObjectMapper();

    record Recorded(String method, String target, String body) {
        @Override
        public String toString() {
            return method + " " + target;
        }
    }

    private final ServerSocket socket;
    private final List<Recorded> requests = new ArrayList<>();
    private volatile String listedSpaceId = DEFAULT_SPACE_ID;

    RecordingHttpServer() throws IOException {
        socket = new ServerSocket(0, 50, InetAddress.getLoopbackAddress());
        Thread accept = new Thread(this::acceptLoop, "recording-http-server");
        accept.setDaemon(true);
        accept.start();
    }

    String url() {
        return "http://127.0.0.1:" + socket.getLocalPort();
    }

    /** The space id the space listing reports for "notes". */
    void listedSpaceId(String id) {
        listedSpaceId = id;
    }

    synchronized List<Recorded> requests() {
        return List.copyOf(requests);
    }

    @Override
    public void close() {
        try { socket.close(); } catch (IOException ignored) {}
    }

    private void acceptLoop() {
        while (!socket.isClosed()) {
            try {
                Socket client = socket.accept();
                Thread handler = new Thread(() -> handle(client));
                handler.setDaemon(true);
                handler.start();
            } catch (IOException closed) {
                return;
            }
        }
    }

    private void handle(Socket client) {
        try (client) {
            InputStream in = client.getInputStream();
            List<String> head = new ArrayList<>();
            for (String line = readLine(in); !line.isEmpty(); line = readLine(in)) head.add(line);
            if (head.isEmpty()) return;

            String[] requestLine = head.get(0).split(" ");
            String length = header(head, "content-length");
            byte[] body = new byte[0];
            if ("chunked".equalsIgnoreCase(header(head, "transfer-encoding"))) {
                ByteArrayOutputStream chunks = new ByteArrayOutputStream();
                for (int size = Integer.parseInt(readLine(in), 16); size > 0;
                     size = Integer.parseInt(readLine(in), 16)) {
                    chunks.write(in.readNBytes(size));
                    readLine(in);
                }
                readLine(in);
                body = chunks.toByteArray();
            } else if (!length.isEmpty()) {
                body = in.readNBytes(Integer.parseInt(length));
            }

            Recorded request = new Recorded(
                    requestLine[0], requestLine[1], new String(body, StandardCharsets.UTF_8));
            synchronized (this) { requests.add(request); }

            Map.Entry<Integer, String> response = respond(request);
            byte[] payload = response.getValue().getBytes(StandardCharsets.UTF_8);
            OutputStream out = client.getOutputStream();
            out.write(("HTTP/1.1 " + response.getKey() + " X\r\n"
                    + "Content-Type: application/json\r\n"
                    + "Content-Length: " + payload.length + "\r\n"
                    + "Connection: close\r\n\r\n").getBytes(StandardCharsets.US_ASCII));
            out.write(payload);
            out.flush();
        } catch (IOException ignored) {
            // The client went away; nothing to record.
        }
    }

    private Map.Entry<Integer, String> respond(Recorded request) throws IOException {
        String route = request.target().split("\\?", 2)[0];
        String method = request.method();

        if (method.equals("GET") && route.equals("/v1/spaces")) {
            var listing = MAPPER.createObjectNode();
            listing.putArray("spaces").addObject().put("name", "notes").put("spaceId", listedSpaceId);
            return Map.entry(200, MAPPER.writeValueAsString(listing));
        }
        if (method.equals("POST") && route.equals("/v1/spaces"))
            return Map.entry(200, "{\"spaceId\":\"" + DEFAULT_SPACE_ID + "\",\"name\":\"notes\"}");
        if (method.equals("POST") && route.equals("/v1/memories:batchGet"))
            return Map.entry(200, "{\"results\":[]}");
        if (method.equals("POST") && route.equals("/v1/memories:retrieve"))
            return Map.entry(200, "");
        if (method.equals("POST") && route.equals("/v1/memories")) {
            JsonNode body = MAPPER.readTree(request.body());
            String id = body.path("memoryId").asText(ASSIGNED_MEMORY_ID);
            return Map.entry(200, MAPPER.writeValueAsString(MAPPER.createObjectNode().put("memoryId", id)));
        }
        if (method.equals("DELETE"))
            return Map.entry(204, "");
        return Map.entry(404, "{\"error\":\"no route\"}");
    }

    private static String header(List<String> head, String name) {
        for (String line : head) {
            int colon = line.indexOf(':');
            if (colon > 0 && line.substring(0, colon).trim().equalsIgnoreCase(name))
                return line.substring(colon + 1).trim();
        }
        return "";
    }

    private static String readLine(InputStream in) throws IOException {
        ByteArrayOutputStream line = new ByteArrayOutputStream();
        for (int b = in.read(); b != -1 && b != '\n'; b = in.read()) {
            if (b != '\r') line.write(b);
        }
        return line.toString(StandardCharsets.ISO_8859_1);
    }
}

File: dotnet/GoodMem.SemanticKernel.Tests/RecordingHttpServer.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json.Nodes;

namespace GoodMem.SemanticKernel.Tests;

/// <summary>
/// A local HTTP server on a real socket. It records the request line of every
/// request exactly as it arrived on the wire, and answers like a minimal
/// GoodMem server with one space named "notes".
/// </summary>
/// <remarks>
/// A raw socket rather than HttpListener or Kestrel, so nothing between the
/// connector and the recording normalises the path.
/// </remarks>
internal sealed class RecordingHttpServer : IDisposable
{
    internal const string DefaultSpaceId = "01a0d16b-bbcd-701c-bfb4-fa306021e078";
    internal const string AssignedMemoryId = "019cfd1d-5a1e-7a4b-9c3e-2f6a1b0c0e09";

    private readonly TcpListener _listener = new(IPAddress.Loopback, 0);
    private readonly CancellationTokenSource _stop = new();
    private readonly object _lock = new();
    private readonly List<RecordedRequest> _requests = [];

    internal sealed record RecordedRequest(string Method, string Target, string Body)
    {
        public override string ToString() => $"{Method} {Target}";
    }

    public RecordingHttpServer()
    {
        _listener.Start();
        _ = Task.Run(AcceptLoopAsync);
    }

    /// <summary>The space id the space listing reports for "notes".</summary>
    public string ListedSpaceId { get; set; } = DefaultSpaceId;

    public string Url => $"http://127.0.0.1:{((IPEndPoint)_listener.LocalEndpoint).Port}";

    public IReadOnlyList<RecordedRequest> Requests
    {
        get { lock (_lock) return _requests.ToList(); }
    }

    public string Describe() => "[" + string.Join(", ", Requests) + "]";

    public void Dispose()
    {
        _stop.Cancel();
        _listener.Stop();
    }

    private async Task AcceptLoopAsync()
    {
        while (!_stop.IsCancellationRequested)
        {
            TcpClient client;
            try { client = await _listener.AcceptTcpClientAsync(_stop.Token).ConfigureAwait(false); }
            catch { return; }
            _ = Task.Run(() => HandleAsync(client));
        }
    }

    private async Task HandleAsync(TcpClient client)
    {
        using var _ = client;
        var stream = client.GetStream();

        // Headers end at the first blank line.
        var lines = new List<string>();
        for (var line = await ReadLineAsync(stream); line is { Length: > 0 }; line = await ReadLineAsync(stream))
            lines.Add(line);
        if (lines.Count == 0) return;

        var requestLine = lines[0].Split(' ');
        string Header(string name) => lines
            .Where(l => l.StartsWith(name + ":", StringComparison.OrdinalIgnoreCase))
            .Select(l => l[(name.Length + 1)..].Trim())
            .FirstOrDefault() ?? "";

        // HttpClient sends JSON content chunked, with no Content-Length.
        var body = new List<byte>();
        if (Header("transfer-encoding").Equals("chunked", StringComparison.OrdinalIgnoreCase))
        {
            for (var size = Convert.ToInt32(await ReadLineAsync(stream), 16); size > 0;
                 size = Convert.ToInt32(await ReadLineAsync(stream), 16))
            {
                body.AddRange(await ReadExactlyAsync(stream, size));
                await ReadLineAsync(stream);
            }
            await ReadLineAsync(stream);
        }
        else if (int.TryParse(Header("content-length"), out var length))
        {
            body.AddRange(await ReadExactlyAsync(stream, length));
        }

        var request = new RecordedRequest(requestLine[0], requestLine[1], Encoding.UTF8.GetString(body.ToArray()));
        lock (_lock) _requests.Add(request);

        var (status, json, contentType) = Respond(request);
        var payload = Encoding.UTF8.GetBytes(json);
        var response =
            $"HTTP/1.1 {status} {(status < 300 ? "OK" : "Not Found")}\r\n" +
            $"Content-Type: {contentType}\r\n" +
            $"Content-Length: {payload.Length}\r\n" +
            "Connection: close\r\n\r\n";
        await stream.WriteAsync(Encoding.ASCII.GetBytes(response)).ConfigureAwait(false);
        await stream.WriteAsync(payload).ConfigureAwait(false);
    }

    private (int Status, string Body, string ContentType) Respond(RecordedRequest request)
    {
        var route = request.Target.Split('?')[0];
        const string Json = "application/json";

        if (request.Method == "GET" && route == "/v1/spaces")
        {
            var space = new JsonObject { ["name"] = "notes", ["spaceId"] = ListedSpaceId };
            return (200, new JsonObject { ["spaces"] = new JsonArray(space) }.ToJsonString(), Json);
        }
        if (request.Method == "POST" && route == "/v1/spaces")
            return (200, new JsonObject { ["spaceId"] = DefaultSpaceId, ["name"] = "notes" }.ToJsonString(), Json);
        if (request.Method == "POST" && route == "/v1/memories:batchGet")
            return (200, """{"results":[]}""", Json);
        if (request.Method == "POST" && route == "/v1/memories:retrieve")
            return (200, "", "application/x-ndjson");
        if (request.Method == "POST" && route == "/v1/memories")
        {
            var memoryId = JsonNode.Parse(request.Body)?["memoryId"]?.GetValue<string>() ?? AssignedMemoryId;
            return (200, new JsonObject { ["memoryId"] = memoryId }.ToJsonString(), Json);
        }
        if (request.Method == "DELETE")
            return (204, "", Json);
        return (404, """{"error":"no route"}""", Json);
    }

    private static async Task<string> ReadLineAsync(NetworkStream stream)
    {
        var bytes = new List<byte>();
        var one = new byte[1];
        while (await stream.ReadAsync(one).ConfigureAwait(false) == 1)
        {
            if (one[0] == '\n') break;
            if (one[0] != '\r') bytes.Add(one[0]);
        }
        return Encoding.Latin1.GetString(bytes.ToArray());
    }

    private static async Task<byte[]> ReadExactlyAsync(NetworkStream stream, int count)
    {
        var buffer = new byte[count];
        await stream.ReadExactlyAsync(buffer).ConfigureAwait(false);
        return buffer;
    }
}
